include/nanobind/stl/filesystem.h

```diff
@@ -0,0 +1,105 @@
+/*
+    nanobind/stl/filesystem.h: type caster for std::string
+
+    Copyright (c) 2022 Qingnan Zhou and Wenzel Jakob
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+#pragma once
+
+#include <nanobind/nanobind.h>
+#include <string>
+
+#ifdef __has_include
+#if __has_include(<filesystem>)
+#include <filesystem>
+#define NB_HAS_FILESYSTEM 1
+#elif __has_include(<experimental/filesystem>)
+#include <experimental/filesystem>
+#define NB_HAS_EXPERIMENTAL_FILESYSTEM 1
+#endif
+#endif
+
+NAMESPACE_BEGIN(NB_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+
+template <typename T>
+struct path_caster {
+private:
+    static PyObject *unicode_from_fs_native(const std::string &w) {
+#if !defined(PYPY_VERSION)
+        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), Py_ssize_t(w.size()));
+#else
+        // PyPy mistakenly declares the first parameter as non-const.
+        return PyUnicode_DecodeFSDefaultAndSize(const_cast<char *>(w.c_str()),
+                                                Py_ssize_t(w.size()));
+#endif
+    }
+
+    static PyObject *unicode_from_fs_native(const std::wstring &w) {
+        return PyUnicode_FromWideChar(w.c_str(), Py_ssize_t(w.size()));
+    }
+
+public:
+    static handle from_cpp(const T &path, rv_policy, cleanup_list *) noexcept {
+        if (auto py_str = unicode_from_fs_native(path.native())) {
+            return module_::import_("pathlib")
+                .attr("Path")(steal(py_str))
+                .release();
+        }
+        return {};
+    }
+
+    bool from_python(handle src, uint8_t, cleanup_list *) noexcept {
+        // PyUnicode_FSConverter and PyUnicode_FSDecoder normally take care of
```

Comment:
The pypy issue mentioned here is ancient, it was created when the PyPy bugtracker was still on Bitbucket.  Can you remove the `PyOS_FSPath` line and actually check that the issue still exists? (We require very recent pypy version in nanobind)

Reply:
That issue seems to be still open: https://foss.heptapod.net/pypy/pypy/-/issues/3168

I am not sure if this is related to the CI failure with pypy. Any suggestions?

Reply:
Yes, I think you are right. That change caused the build error that now appears on PyPy, so it still does seem to be an issue. (That issue was so stale that I assumed it might have been fixed in the meantime without updating the issue tracker ticket)